File: 278.first-bad-version.py
#
# @lc app=leetcode id=278 lang=python3
#
# [278] First Bad Version
#
from math import ceil
# @lc code=start
# The isBadVersion API is already defined for you.
# @param version, an integer
# @return an integer
# def isBadVersion(version):

class Solution:
    def firstBadVersion(self, n: int) -> int:
        m = 1
        oldest_bad_version = n
        while m <= n:
            target = (m + n) // 2
            if isBadVersion(target):
                oldest_bad_version = target
                n = target - 1
            else:
                m = target + 1
            
        return oldest_bad_version

# @lc code=end

File: 278.first-bad-version.cpp
/*
 * @lc app=leetcode id=278 lang=cpp
 *
 * [278] First Bad Version
 */

// The API isBadVersion is defined for you.
bool isBadVersion(int version);
// @lc code=start

class Solution {
public:
  int firstBadVersion(int n) {
    int answer;
    unsigned int i = 1;
    unsigned int j = n;
    unsigned int target_idx;
    while (i <= j) {
      target_idx = (i + j) / 2;

      if (isBadVersion(target_idx)) {
        answer = target_idx;
        j = target_idx - 1;
      } else {
        i = target_idx + 1;
      }
    }

    return answer;
  }
};
// @lc code=end

